### PR TITLE
New: Nested Tabs - Use the old tabs if the experiment is Off [ED-7325] [ED-7326]

### PR DESCRIPTION
**Pull request by iNewLegend**

### CODE REVIEW

File: assets/dev/js/editor/regions/panel/pages/elements/elements.js [hunk 1]
```diff
@@ -65,6 +65,14 @@ PanelElementsLayoutView = Marionette.LayoutView.extend( {
 		const elementsCollection = new PanelElementsElementsCollection(),
 			isContainerActive = elementorCommon.config.experimentalFeatures.container;
 
+		Object.entries( elementor.widgetsCache ).forEach( ( [ widgetName, widgetData ] ) => {
+			// If new version loaded.
```
Comment (TzviRabinovitch) [inline assets/dev/js/editor/regions/panel/pages/elements/elements.js:69]: remove comment 

please write this code in a more clean way  :

the final code should look like 
Object.entries( elementor.widgetsCache ).forEach(hideOldTabWidgetsIfNeeded(paramsNeeded))

hideOldTabWidgetsIfNeeded should look like 
hideOldTabWidgetsIfNeeded (paramsNeeded){
  if (isNewTabWidgetLoaded){
     hideOldTabWidget();
   }
}

hideOldTabWidget  should look like 
hideOldTabWidget(paramsNeeded )
elementor.widgetsCache[ widgetName ].show_in_panel = false
]

File: assets/dev/js/editor/regions/panel/pages/elements/elements.js [hunk 2]
```diff
@@ -65,6 +65,14 @@ PanelElementsLayoutView = Marionette.LayoutView.extend( {
...
+			if ( widgetData.mark_old_version_of && elementor.widgetsCache[ widgetData.mark_old_version_of ] ) {
+				// Hide the old version.
```
Comment (TzviRabinovitch) [inline assets/dev/js/editor/regions/panel/pages/elements/elements.js:71]: remove comment after refactoring

File: modules/tabs-v2/module.php [hunk 3]
```diff
@@ -13,7 +13,7 @@ class Module extends \Elementor\Core\Base\Module {
 	public static function get_experimental_data() {
 		return [
 			'name' => 'tabs-v2',
-			'title' => esc_html__( 'Nested Tab', 'elementor' ),
+			'title' => esc_html__( 'Nested Tabs', 'elementor' ),
 			'description' => esc_html__( 'Nested Tabs', 'elementor' ), // TODO: Add description
```
Comment (TzviRabinovitch) [inline modules/tabs-v2/module.php:17]: add description please 
don't push code with TODO

**Review by TzviRabinovitch — CHANGES_REQUESTED** [review]

File: assets/dev/js/editor/regions/panel/pages/elements/elements.js (hunk 1 above)
Comment (matipojo) [reply assets/dev/js/editor/regions/panel/pages/elements/elements.js:69]: Hi @TzviRabinovitch it should be a generic code that will support other deprecated widgets

**Review by matipojo — CHANGES_REQUESTED** [review]

File: assets/dev/js/editor/regions/panel/pages/elements/elements.js (hunk 1 above)
Comment (TzviRabinovitch) [reply assets/dev/js/editor/regions/panel/pages/elements/elements.js:69]: @matipojo then write it generic 
you can add the widget name to hide or any other param in the call to the function 
clean code can be generic too :)

File: includes/widgets/tabs.php [hunk 4]
```diff
@@ -74,6 +74,14 @@ public function get_keywords() {
 		return [ 'tabs', 'accordion', 'toggle' ];
 	}
 
+	protected function get_initial_config() {
+		$config = parent::get_initial_config();
+
+		$config['mark_old_version_of'] = 'tabs-v2';
```
Comment (matipojo) [inline includes/widgets/tabs.php:80]:
```suggestion
		$config['deprecated'] = [
			'version' => '3.8.0',
			'replacement' => 'tabs-v2',
		];
```

**Review by matipojo — CHANGES_REQUESTED** [review]

File: includes/widgets/tabs.php (hunk 4 above)
Comment (iNewLegend) [reply includes/widgets/tabs.php]: Handled.

**Review by matipojo — APPROVED** [review]

**Review by matipojo — CHANGES_REQUESTED** [review]
Please cover the new `deprecated` prop with tests

File: includes/widgets/tabs.php [hunk 5]
```diff
@@ -74,6 +74,17 @@ public function get_keywords() {
 		return [ 'tabs', 'accordion', 'toggle' ];
 	}
 
+	protected function get_initial_config() {
+		$config = parent::get_initial_config();
+
+		$config['deprecated'] = [
```
Comment (iNewLegend) [inline includes/widgets/tabs.php:80]: @matipojo it would be nice if we cloud have document for all widget config options.

Comment (matipojo) [reply includes/widgets/tabs.php:80]: Extract to a method `mark_as_deprecated( $config )` on the `Contorls_Stack` so it will be a part of the infra.

Comment (matipojo) [reply includes/widgets/tabs.php:80]: Also, mark it as deprecated only if the new experiment is active

File: includes/widgets/tabs.php [hunk 6]
```diff
@@ -117,6 +128,19 @@ protected function register_controls() {
 			]
 		);
 
+		// Check if TabsV2 experiment is enabled.
+		if ( Plugin::$instance->experiments->is_feature_active( 'tabs-v2' ) ) {
```
Comment (matipojo) [inline includes/widgets/tabs.php:132]: Make it generic, add a method `deprecation_message` that is based on the `deprecated` config.
```suggestion
		$this->deprecation_message();
```

**Review by matipojo — CHANGES_REQUESTED** [review]

File: tests/playwright/sanity/includes/widgets/tabs.test.js [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+const WpAdminPage = require( '../../../pages/wp-admin-page.js' );
+
+test( 'Ensure the old tabs widget is telling deprecation warning message', async ( { page }, testInfo ) => {
+	// Arrange.
+	const wpAdmin = new WpAdminPage( page, testInfo );
+
+	await wpAdmin.setExperiments( {
+		container: true,
+		'nested-elements': true,
+		'tabs-v2': true,
```
Comment (iNewLegend) [inline tests/playwright/sanity/includes/widgets/tabs.test.js:11]:
```suggestion

```

File: includes/base/widget-base.php [hunk 8]
```diff
@@ -3,6 +3,8 @@
 
 use Elementor\Core\Page_Assets\Data_Managers\Responsive_Widgets as Responsive_Widgets_Data_Manager;
 use Elementor\Core\Page_Assets\Data_Managers\Widgets_Css as Widgets_Css_Data_Manager;
+use Elementor\Modules\DevTools\Deprecation;
+use WPForms\Integrations\Elementor\Elementor;
```
Comment (iNewLegend) [inline includes/base/widget-base.php:7]:
```suggestion
```

**Review by matipojo — CHANGES_REQUESTED** [review]
Please resolve the conflicts

File: includes/base/widget-base.php [hunk 9]
```diff
@@ -996,6 +1002,32 @@ public function get_inline_css_depends() {
 		return [];
 	}
 
+	/**
```
Comment (matipojo) [inline includes/base/widget-base.php:1005]: Please move the new protected methods after the public.

File: modules/tabs-v2/module.php [hunk 10]
```diff
@@ -9,7 +9,6 @@
 
 class Module extends \Elementor\Core\Base\Module {
 
-	public static function is_active() {
```
Comment (matipojo) [inline modules/tabs-v2/module.php:12]: ?

**Review by matipojo — CHANGES_REQUESTED** [review]
Please fix the checks errors

File: includes/base/widget-base.php [hunk 11]
```diff
@@ -1075,6 +1080,44 @@ public function get_responsive_widgets() {
 		return $responsive_widgets_data_manager->get_asset_data_from_config( $config );
 	}
 
+	/**
+	 * Mark widget as deprecated.
+	 *
+	 * The return should have 3 elements:
```
Comment (iNewLegend) [inline includes/base/widget-base.php:1086]:
```suggestion
	 * The return should have thw following config:
```

**Review by Nevoss — CHANGES_REQUESTED** [review]

File: includes/base/widget-base.php [hunk 12]
```diff
@@ -1075,6 +1074,39 @@ public function get_responsive_widgets() {
...
+	 * Mark widget as deprecated.
+	 *
+	 * Use `get_deprecation_message()` method to print the message control at specific location in register_controls().
+	 *
+	 * @param $version string		The version of Elementor that deprecated the widget.
+	 * @param $message string 		A message regarding the deprecation.
+	 * @param $replacement string	The widget that should be used instead.
+	 */
+	protected function add_deprecation_message( $version, $message, $replacement ) {
+		// Expose the config for handling in JS
```
Comment (matipojo) [inline includes/base/widget-base.php:1087]:
```suggestion
		// Expose the config for handling in JS.
```

File: includes/base/widget-base.php [hunk 13]
```diff
@@ -1075,6 +1074,39 @@ public function get_responsive_widgets() {
...
 	}
 
+	/**
+	 * Mark widget as deprecated.
+	 *
+	 * Use `get_deprecation_message()` method to print the message control at specific location in register_controls().
+	 *
+	 * @param $version string		The version of Elementor that deprecated the widget.
+	 * @param $message string 		A message regarding the deprecation.
+	 * @param $replacement string	The widget that should be used instead.
```
Comment (matipojo) [inline includes/base/widget-base.php:1084]:
```suggestion
	 * @param $version string           The version of Elementor that deprecated the widget.
	 * @param $message string         A message regarding the deprecation.
	 * @param $replacement string   The widget that should be used instead.
```

File: includes/base/widget-base.php [hunk 14]
```diff
@@ -1075,6 +1074,39 @@ public function get_responsive_widgets() {
...
+	 * @param $message string         A message regarding the deprecation.
+	 * @param $replacement string   The widget that should be used instead.
+	 */
+	protected function add_deprecation_message( $version, $message, $replacement ) {
+		// Expose the config for handling in JS.
+		$this->set_config( 'deprecation', [
+			'version' => $version,
+			'message' => $message,
+			'replacement' => $replacement,
+		] );
```
Comment (Nevoss) [inline includes/base/widget-base.php:1092]: Should we handle js deprecation if the replacement is not active?

**Review by Nevoss — CHANGES_REQUESTED** [review]

File: includes/base/widget-base.php (hunk 14 above)
Comment (matipojo) [reply includes/base/widget-base.php:1092]: No, but it shouldn't be a part of this method, fixed.

**Review by Nevoss — DISMISSED** [review]

**Pull request merged**